### ISSUE TITLE
Define persistence policy

### ISSUE THREAD

**Issue by jvalkeal** [spring-projects/spring-statemachine#430]
Now that we're trying to implement easier way for user to persist a machine, and as this first iteration is going to be based on interceptor it's probably a good idea to come up with a policy of some sort. Maybe user want to persist before/after state has been set or define if persisting failure should prevent state change to happen, etc. This would align with ideas of not force feeding some specific functionality. Also should think about introducing more meaningful interface around these ideas.

**Comment by burakhelvaci** [comment]
Do you mind, adding reactive support for StateMachineInterceptor,(maybe StateMachineListener and StateMachineMonitor). It can be required. Especially if user want to persist state, stateContext etc.. in StateMachineInterceptor. 

Actually I'm tried to contribute for reactive supported StateMachineInterceptor but reactive changes effects lots of things that I cannot handle with. So I give up 😃

**Comment by jvalkeal** [comment]
Yeah that is a good point. Not that much value right now but it's easier to do api changes now. I'll see if I can revamp that interface for next milestone.
